### ISSUE TITLE
`re.compile` receiving generically typed argument gives error

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5794]
### Discussed in https://github.com/microsoft/pyright/discussions/5793

<div type='discussions-op-text'>

<sup>Originally posted by **pvandyken** August 23, 2023</sup>
I'm trying to understand why pyright reports an error on the following code:

```python
from __future__ import annotations
import re
from typing import AnyStr

def compile(a: AnyStr | re.Pattern[AnyStr]):
    re.compile(a)
	#		   ~
    # Argument of type "AnyStr@compile | Pattern[AnyStr@compile]" cannot be assigned to parameter "pattern" of type "Pattern[AnyStr@compile]" in function "compile"
	#  Type "str* | bytes* | Pattern[AnyStr@compile]" cannot be assigned to type "Pattern[AnyStr@compile]"
	#    "bytes*" is incompatible with "Pattern[AnyStr@compile]"
```

Since `re.compile` should take either a `AnyStr` or `Pattern` and return a `Pattern` of the same generic type, I wouldn't expect this to have any problem. Does this actually represent unsafe code, or is this a technical limitation in pyright?</div>

**Comment by erictraut** [comment]
Here's a related issue:

```python
from re import Pattern
from typing import Any, AnyStr, overload

@overload
def compile(pattern: AnyStr) -> Pattern[AnyStr]:
    ...

@overload
def compile(pattern: Pattern[AnyStr]) -> Pattern[AnyStr]:
    ...

def compile(pattern: Any) -> Any:
    return

def func1(a: str | bytes) -> Pattern[str] | Pattern[bytes]:
    return compile(a)  # False positive

def func2(a: AnyStr) -> Pattern[AnyStr]:
    return compile(a)
```

The above type checks without error in mypy.

Interestingly, the following simplified example does not type check in mypy.

```python
from typing import AnyStr

def compile(pattern: AnyStr) -> AnyStr:
    return pattern

def func1(a: str | bytes) -> str | bytes:
    return compile(a)
```

But adding a dummy overload allows it to type check. That's because mypy (and pyright) apply union expansion for overloads but not for simple (non-overloaded) functions.

```python
from typing import Any, AnyStr, overload

@overload
def compile(pattern: AnyStr) -> AnyStr:
    ...

@overload
def compile(pattern: int) -> None:
    ...

def compile(pattern: Any) -> Any:
    return

def func1(a: str | bytes) -> str | bytes:
    return compile(a)
```

**Comment by erictraut** [comment]
This is included in pyright 1.1.325, which I just published. It will also be included in an upcoming pylance release.
